virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,219 @@
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/nvmem-provider.h>
+#include <linux/slab.h>
+#include <linux/mutex.h>
+#include <linux/of.h>
+#include <linux/list.h>
+#include <linux/fs.h>
+#include <linux/fcntl.h>
+#include <linux/err.h>
+#include <linux/uaccess.h>
+#include <linux/platform_device.h>
+
+// Structure to hold EEPROM data and metadata
+struct virtual_eeprom {
+    char *data;
+    struct nvmem_device *nvmem_dev;
+    struct nvmem_config nvmem_cfg;
+    struct mutex mutex;
+    int size;
+    const char *name;
+    const char *bin_path;
+    struct list_head list;
+    struct platform_device *pdev;
+};
+
+// Function prototypes
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom);
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom);
+
+// Initialize global list to track all EEPROMs
+static LIST_HEAD(eeprom_list);
+
+static int eeprom_read(void *context, unsigned int offset, void *val, size_t bytes)
+{
+    struct virtual_eeprom *eeprom = context;
+
+    if (offset >= eeprom->size)
+        return -EINVAL;
+
+    if (offset + bytes > eeprom->size)
```

Comment:
If it cannot read all the bytes requested it should return an error, there is no allowance for a partial read.

Reply:
same with write

Reply:
fixed